### PR TITLE
src: add Long command descriptions for each of the CLI commands

### PR DESCRIPTION
**Pull request by lance**
Uses the Cobra "Long" configuration for each command to provide more
descriptive text.

Example:

```console
faas help create                                                                                                      1.3m  Mon 21 Sep 2020 09:55:40 PM EDT
Create a new Function, including initialization of local files and deployment

Creates a new Function project at 'path'. If 'path' does not exist, it is
created. The function name is the name of the leaf directory at path. After
creating the project, a container image is created and is deployed. This
command wraps 'init', 'build' and 'deploy' all up into one command.

The runtime, trigger, image name, image repository, and namespace may all be
specified as flags on the command line, and will subsequently be the default
values when an image is built or a Function is deployed. If the image name and
image repository are both unspecified, the user will be prompted for a
repository name, and the image name can be inferred from that plus the function
name. The function name, namespace, image name and repository name are all
persisted in the project configuration file .faas.yaml.

Usage:
  faas create <name> [options] [flags]

Flags:
  -c, --confirm             Prompt to confirm all configuration options - $FAAS_CONFIRM
  -h, --help                help for create
  -i, --image string        Optional full image name, in form [registry]/[namespace]/[name]:[tag] for example quay.io/myrepo/project.name:latest (overrides --repository) - $FAAS_IMAGE
  -n, --namespace string    Override namespace into which the Function is deployed (on supported platforms).  Default is to use currently active underlying platform setting - $FAAS_NAMESPACE
  -p, --path string         Path to the new project directory - $FAAS_PATH (default "/home/lanceball/src/go/src/github.com/boson-project/faas")
  -r, --repository string   Repository for built images, ex 'docker.io/myuser' or just 'myuser'.  Optional if --image provided. - $FAAS_REPOSITORY
  -l, --runtime string      Function runtime language/framework. - $FAAS_RUNTIME (default "go")
      --templates string    Extensible templates path. - $FAAS_TEMPLATES (default "/home/lanceball/.config/faas/templates")
  -t, --trigger string      Function trigger (ex: 'http','events') - $FAAS_TRIGGER (default "http")

Global Flags:
      --config string   config file path (default "~/.faas/config")
  -v, --verbose         print verbose logs
```

### CODE REVIEW

File: cmd/create.go [hunk 1]
```diff
@@ -22,9 +22,9 @@ func init() {
 	createCmd.Flags().StringP("namespace", "n", "", "Override namespace into which the Function is deployed (on supported platforms).  Default is to use currently active underlying platform setting - $FAAS_NAMESPACE")
 	createCmd.Flags().StringP("path", "p", cwd(), "Path to the new project directory - $FAAS_PATH")
 	createCmd.Flags().StringP("repository", "r", "", "Repository for built images, ex 'docker.io/myuser' or just 'myuser'.  Optional if --image provided. - $FAAS_REPOSITORY")
-	createCmd.Flags().StringP("runtime", "l", faas.DefaultRuntime, "Function runtime language/framework. - $FAAS_RUNTIME")
+	createCmd.Flags().StringP("runtime", "l", faas.DefaultRuntime, "Function runtime language/framework. Default available runtimes: 'node', 'quarkus' and 'go'. - $FAAS_RUNTIME")
```
Comment (zroubalik) [inline cmd/create.go:25]: This should be imho split to `Default runtime` and `Available runtimes:`. From the description I am not able to say which runtim is the default one.

File: cmd/create.go [hunk 2]
```diff
@@ -22,9 +22,9 @@ func init() {
...
 	createCmd.Flags().StringP("templates", "", filepath.Join(configPath(), "templates"), "Extensible templates path. - $FAAS_TEMPLATES")
-	createCmd.Flags().StringP("trigger", "t", faas.DefaultTrigger, "Function trigger (ex: 'http','events') - $FAAS_TRIGGER")
+	createCmd.Flags().StringP("trigger", "t", faas.DefaultTrigger, "Function trigger. Available triggers: 'http' and 'events' - $FAAS_TRIGGER")
```
Comment (zroubalik) [inline cmd/create.go:27]: Which trigger is default?

File: cmd/delete.go [hunk 3]
```diff
@@ -17,9 +17,19 @@ func init() {
...
 
 var deleteCmd = &cobra.Command{
-	Use:               "delete <name>",
-	Short:             "Delete a Function deployment",
-	Long:              `Removes the deployed Function by name, by explicit path, or by default for the current directory.  No local files are deleted.`,
+	Use:   "delete <name>",
+	Short: "Delete a Function deployment",
+	Long: `Delete a Function deployment
+Removes a deployed function from the cluster. The user may specify a function
+by name, path or if neither of those are provided, the current directory will
```
Comment (zroubalik) [inline cmd/delete.go:24]:
```suggestion
by name, path using the --path or -p flag or if neither of those are provided, the current directory will
```

File: cmd/init.go [hunk 4]
```diff
@@ -15,18 +15,32 @@ func init() {
 	root.AddCommand(initCmd)
 	initCmd.Flags().BoolP("confirm", "c", false, "Prompt to confirm all configuration options - $FAAS_CONFIRM")
 	initCmd.Flags().StringP("path", "p", cwd(), "Path to the new project directory - $FAAS_PATH")
-	initCmd.Flags().StringP("runtime", "l", faas.DefaultRuntime, "Function runtime language/framework. - $FAAS_RUNTIME")
+	initCmd.Flags().StringP("runtime", "l", faas.DefaultRuntime, "Function runtime language/framework. Default available runtimes: 'node', 'quarkus' and 'go'. - $FAAS_RUNTIME")
```
Comment (zroubalik) [inline cmd/init.go:18]: see `create` for remark on default/available runtimes

File: cmd/init.go [hunk 5]
```diff
@@ -15,18 +15,32 @@ func init() {
...
 	initCmd.Flags().StringP("templates", "", filepath.Join(configPath(), "templates"), "Extensible templates path. - $FAAS_TEMPLATES")
-	initCmd.Flags().StringP("trigger", "t", faas.DefaultTrigger, "Function trigger (ex: 'http','events') - $FAAS_TRIGGER")
+	initCmd.Flags().StringP("trigger", "t", faas.DefaultTrigger, "Function trigger. Available triggers: 'http' and 'events' - $FAAS_TRIGGER")
```
Comment (zroubalik) [inline cmd/init.go:20]: see `create` for remark on default trigger

File: cmd/create.go [hunk 6]
```diff
@@ -38,8 +38,23 @@ func init() {
 }
 
 var createCmd = &cobra.Command{
-	Use:        "create <name>",
-	Short:      "Create a new Function, including initialization of local files and deployment.",
+	Use:   "create <name>",
+	Short: "Create a new Function, including initialization of local files and deployment",
+	Long: `Create a new Function, including initialization of local files and deployment
+
+Creates a new Function project at 'path'. If 'path' does not exist, it is
```
Comment (zroubalik) [inline cmd/create.go:45]: We are talking about 'path' in here, but it is not clear, where in the command the path should be specified.

File: cmd/init.go [hunk 7]
```diff
@@ -15,18 +15,32 @@ func init() {
...
 }
 
 var initCmd = &cobra.Command{
-	Use:        "init <name>",
-	Short:      "Initialize a new Function project",
+	Use:   "init <name>",
+	Short: "Initialize a new Function project",
+	Long: `Initializes a new Function project
+
+Creates a new Function project at 'path'. If 'path' is unspecified, assumes the
```
Comment (zroubalik) [inline cmd/init.go:32]: same remark on 'path' as mentioned in `create`. And this description is missing the part on the function name inherited from the leaf directory.

**Review by zroubalik — COMMENTED** [review]
Looking good, minor nits

**Comment by lance** [comment]
@zroubalik thanks for your feedback. I've incorporated it.

File: cmd/init.go [hunk 8]
```diff
@@ -15,18 +15,31 @@ func init() {
...
 
 	if err := initCmd.RegisterFlagCompletionFunc("runtime", CompleteRuntimeList); err != nil {
 		fmt.Println("Error while calling RegisterFlagCompletionFunc: ", err)
 	}
 }
 
 var initCmd = &cobra.Command{
-	Use:        "init <name>",
-	Short:      "Initialize a new Function project",
+	Use:   "init <name>",
```
Comment (zroubalik) [inline cmd/init.go:28]: minor nit, there should be `init <path>`

**Review by zroubalik — APPROVED** [review]
LGTM, minor nit

File: cmd/init.go (hunk 8 above)
Comment (lance) [reply cmd/init.go]: Oops!

**Comment by nainaz** [comment]
This looks good.

**Review by rhuss — COMMENTED** [review]
Thanks a lot, I added my comments below. But let me sum the up here

*  `create` use an path as argument, all other commands use an option `--path`. This should be similar.
* I doubt the usefulness of `create` except for demos or simple smoketest if the user as no chance to modify the code between scaffolding and building. Also, as this command will also automatically push to a registry this might add to 'image pollution' (there is no easy way to "unpush")
* I would make `.faas.yml` more explicit,  much like in build systems (npm, maven) or also serverless.com (which uses a serverless.json). This would have the additional benefit that this pattern is known to most developers already. Also, this would allow you to focus on this file (which contains the meta data) instead of mangling around with directory hierachies.

I could imagine a usage pattern like in these hypothetical examples:

```
# Scaffold a project in a subdirectory "helloworld" (to be created) and create a faas.yml file within it
faas init helloworld
cd helloworld

# Same as above but using the special name "." which creates the faas.yml in the local directory and picks up the name from the parent directory
mkdir helloworld
cd helloworld
faas init .

# Build the image with the project from the current directory
faas build

# Build the image from another faas.yml, pick up all meta data from there
faas build -f otherfunction/faas.yml

# Build the image from another directory that must contain a faas.yml
fass build -f /myother/dir

# Same semantics for all other commands ...
```

This pattern is user by Maven and (I believe, please correct me) by npm as well.

File: cmd/build.go [hunk 9]
```diff
@@ -20,8 +20,20 @@ func init() {
...
 
 var buildCmd = &cobra.Command{
-	Use:        "build",
-	Short:      "Build an existing Function project as an OCI image",
+	Use:   "build",
+	Short: "Build an existing Function project as an OCI image",
+	Long: `Build an existing Function project as an OCI image
+
+Builds the Function project in the current directory or in the directory
+specified by the --path flag. The .faas.yaml file is read to determine the
```
[the remark this answers is not in the document]
Comment (lance) [reply cmd/build.go:28]: You make a good point here.

File: cmd/create.go [hunk 10]
```diff
@@ -38,8 +38,23 @@ func init() {
...
 var createCmd = &cobra.Command{
-	Use:        "create <name>",
-	Short:      "Create a new Function, including initialization of local files and deployment.",
+	Use:   "create <path>",
+	Short: "Create a new Function, including initialization of local files and deployment",
+	Long: `Create a new Function, including initialization of local files and deployment
+
+Creates a new Function project at <path>. If <path> does not exist, it is
+created. The Function name is the name of the leaf directory at <path>. After
+creating the project, a container image is created and is deployed. This
```
[the remark this answers is not in the document]
Comment (lance) [reply cmd/create.go:47]: Yes, <path> is mandatory here and may be `.` - at least when [this](https://github.com/boson-project/faas/pull/127/files) lands.

File: cmd/create.go [hunk 11]
```diff
@@ -38,8 +38,23 @@ func init() {
...
+command wraps "init", "build" and "deploy" all up into one command.
+
```
[the remark this answers is not in the document]
Comment (lance) [reply cmd/create.go:49]: > tbh, I question the usefulness of a faas create that does scaffolding and then an intermediate build afterwards

I am with you on this. I believe @zroubalik and @matejvasek are as well. But we'll have to convince @lkingland and @nainaz.

> Instead I would love to see non-overlapping commands, that do:
>   * Scaffolding (init or maybe create but only for creating the repo)
>   * Building (build)
>   * Deploy (deploy)
>   * Run locally (run)

This already exists in what we have today. It's just that there is an additional `create` command.

File: cmd/create.go [hunk 12]
```diff
@@ -38,8 +38,23 @@ func init() {
...
+The runtime, trigger, image name, image repository, and namespace may all be
+specified as flags on the command line, and will subsequently be the default
+values when an image is built or a Function is deployed. If the image name and
+image repository are both unspecified, the user will be prompted for a
+repository name, and the image name can be inferred from that plus the function
```
[the remark this answers is not in the document]
Comment (lance) [reply cmd/create.go:54]: I know! It's super frustrating. At one point, early on in this project, I was using `--tag` similar to Docker, which considers the entire string of `registry/user/repository:tag` to be the `--tag`. But we reverted.

As it stands now, "image name" specified by `--image` is the full string `registry/user/repository:tag`, and repository is `registry/user`. I would __love__ to normalize this usage based on existing tooling in the wild, but it's not really clear how to best go about that given that each registry may have different URLs. For example, Docker supports `docker.io/node:latest` (no user in this case).

File: cmd/delete.go [hunk 13]
```diff
@@ -17,9 +17,20 @@ func init() {
...
-	Long:              `Removes the deployed Function by name, by explicit path, or by default for the current directory.  No local files are deleted.`,
+	Use:   "delete <name>",
+	Short: "Delete a Function deployment",
+	Long: `Delete a Function deployment
+
+Removes a deployed function from the cluster. The user may specify a function
+by name, path using the --path or -p flag, or if neither of those are provided,
+the current directory will be searched for a .faas.yaml configuration file to
+determine the function to be removed.
+
```
[the remark this answers is not in the document]
Comment (lance) [reply cmd/delete.go:28]: If you have some time to read the history, I recently [introduced a PR](https://github.com/boson-project/faas/pull/125) that removed `--path` altogether from almost all commands, leaving it only for `init` and `create` as a required argument. But that PR was closed as seeming [less scriptable](https://github.com/boson-project/faas/pull/125#issuecomment-694843551).

I am happy to reconsider this. My thinking was that if all commands are executed from within the context of a project directory, there is no need for a `--path` argument except when creating a project. Given that the PR was reverted, all commands now have `--path` as a flag, and `init` and `create` have <path> as a required argument. I wouldn't mind seeing `--path` go away, but I'd prefer not to make <path> in `create` and `init` a required flag. Maybe it's irrational, but I just don't like required flags. :smiley:

File: cmd/create.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (zroubalik) [reply cmd/create.go:49]: and the deploy/update should be merged

[the remark this answers is not in the document]
Comment (zroubalik) [reply cmd/create.go:49]: > tbh, I question the usefulness of a `faas create` that does scaffolding and then an intermediate build afterwards:
> 
>     * The user has no way to inspect the code that she is automatically publishing at the Docker registry
> 
>     * Its just useful for demo purposes only to deploy a sample code that can't be changed by the user.
> 
>     * `create` can be only called once, so the benefit of combining multiple commands is minimal.
> 
> 
> Instead I would love to see non-overlapping commands, that do:
> 
>     * Scaffolding (`init` or maybe `create` but only for creating the repo)
> 
>     * Building (`build`)
> 
>     * Deploy (`deploy`)
> 
>     * Run locally (`run`)
+1

File: cmd/build.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (nainaz) [reply cmd/build.go:28]: Agreed. visible file would help the Fucntion Authors those are not from developers background.

File: cmd/create.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (nainaz) [reply cmd/create.go:48]: Personally, I like individual commands. so no need to convince me there but trying to see it from the perspective of people who don't have developer background makes me think that "create" might be a great hook. 
If we do end up deleting "create" as a command. I would say, we should rename init -> create.

**Pull request merged**